### PR TITLE
DPC-1560 post and delete token

### PR DESCRIPTION
**Pull request by MrBilnon**
### Fixes [DPC-1560](https://jira.cms.gov/browse/DPC-1560)

In SSAS v2, we need to be able to create and delete (multiple) client tokens in a system.

### Change Details

Added two methods, one to delete and one to create, the create method uses an existing method from when we registered a v2 system and the delete method uses gorm to retrieve the root key by token uuid and system id and then deletes both the root key and client token associated with that root key.  All deletes are soft deletes.

### Security Implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### Feedback Requested

* Should there be a label length limit?

### CODE REVIEW

File: ssas/service/admin/api.go [hunk 1]
```diff
@@ -663,6 +663,99 @@ func deleteSystemIP(w http.ResponseWriter, r *http.Request) {
...
+
+	if err := json.Unmarshal(b, &body); err != nil {
+		ssas.OperationFailed(tokenEvent)
+		jsonError(w, http.StatusInternalServerError, "Internal Error")
+		return
+	}
+
+	if body["label"] == "" {
+		ssas.OperationFailed(tokenEvent)
+		jsonError(w, http.StatusInternalServerError, "Internal Error")
```
Comment (dhgreene) [inline ssas/service/admin/api.go:703]: It doesn't feel like an internal error if required input is missing

File: ssas/service/admin/api.go [hunk 2]
```diff
@@ -663,6 +663,99 @@ func deleteSystemIP(w http.ResponseWriter, r *http.Request) {
...
+	response := ssas.ClientTokenResponse{
+		ClientTokenOutput: ssas.OutputCT(*ct)[0],
+		Token:             m,
+	}
+
+	b, err = json.Marshal(response)
+	if err != nil {
+		tokenEvent.Help = fmt.Sprintf("could not marshal response for clientID %s, groupID %s: %s", system.ClientID, system.GroupID, err.Error())
+		ssas.OperationFailed(tokenEvent)
+		ssas.OperationFailed(tokenEvent)
```
Comment (dhgreene) [inline ssas/service/admin/api.go:725]:
```suggestion
```

File: ssas/service/admin/api.go [hunk 3]
```diff
@@ -663,6 +663,99 @@ func deleteSystemIP(w http.ResponseWriter, r *http.Request) {
...
+		jsonError(w, http.StatusInternalServerError, "Internal Error")
+		return
+	}
+
+	expiration := time.Now().Add(ssas.MacaroonExpiration)
+	ct, m, err := system.SaveClientToken(body["label"], group.XData, expiration)
+	if err != nil {
+		tokenEvent.Help = fmt.Sprintf("could not save client token for clientID %s, groupID %s: %s", system.ClientID, system.GroupID, err.Error())
+		ssas.OperationFailed(tokenEvent)
+		ssas.OperationFailed(tokenEvent)
```
Comment (dhgreene) [inline ssas/service/admin/api.go:712]:
```suggestion
```

File: ssas/service/admin/api.go [hunk 4]
```diff
@@ -663,6 +663,99 @@ func deleteSystemIP(w http.ResponseWriter, r *http.Request) {
...
+
+	system, err := ssas.GetSystemByID(systemID)
+	if err != nil {
+		ssas.OperationFailed(tokenEvent)
+		jsonError(w, http.StatusNotFound, "Invalid system ID")
+		return
+	}
+
+	ssas.OperationCalled(tokenEvent)
+	err = system.DeleteClientToken(tokenID)
```
Comment (dhgreene) [inline ssas/service/admin/api.go:749]: Should we verify that the token belongs to the system?

File: ssas/systems.go [hunk 5]
```diff
@@ -135,6 +135,28 @@ func (system *System) GetClientTokens(trackingID string) ([]ClientToken, error)
...
+func (system *System) DeleteClientToken(tokenID string) error {
+	db := GetGORMDbConnection()
+	defer Close(db)
+
+	tx := db.Begin()
+	defer func() {
+		if r := recover(); r != nil {
+			tx.Rollback()
+		}
+	}()
```
Comment (dhgreene) [inline ssas/systems.go:147]: When there's just one SQL statement, do we gain anything by using a transaction?

Comment (MrBilnon) [reply ssas/systems.go:147]: @dhgreene it's two sql statements though

Comment (dhgreene) [reply ssas/systems.go:147]: But it doesn't have to be 😈

Comment (dhgreene) [reply ssas/systems.go:147]: On third reading, I see that you're wanting to delete both the root key and the client token. 👍

**Review by dhgreene — APPROVED** [review]
🦖

**Pull request merged**